content/curried-varargs.md

```diff
@@ -0,0 +1,271 @@
+      @inline def apply(attributesAndChildren: Any*) = {
+        attributesAndChildren.foreach {
+          case text: String =>
+            element.appendChild(document.createTextNode(text))
+          case node: Node =>
+            element.appendChild(node)
+          case attribute: Attribute[E] =>
+            attribute(element)
+        }
+        element
+      }
+    }
+    @inline def div = new Builder(document.createElement("div"))
+    @inline def br = new Builder(document.createElement("br"))
+  }
+}
+```
+
+However, the Scala compiler is unable to optimize repeated parameters, as a result, the output JavaScript from Scala.js would look like the below code.
+
+``` javascript
+var $$this$1 = $m_Lorg_scalajs_dom_package$().document__Lorg_scalajs_dom_raw_HTMLDocument().createElement("div");
+var this$3 = $m_LScalaFiddle$xml$attributes$();
+var jsx$1 = new $c_sjsr_AnonFunction1().init___sjs_js_Function1((function($this, value) {
+  return (function(x$1$2) {
+    x$1$2.setAttribute("title", value)
+  })
+})(this$3, "my-title"));
+var $$this = $m_Lorg_scalajs_dom_package$().document__Lorg_scalajs_dom_raw_HTMLDocument().createElement("br");
+var array = [jsx$1, "line1", $$this, "line2"];
+var i = 0;
+var len = $uI(array.length);
+while ((i < len)) {
+  var index = i;
+  var arg1 = array[index];
+  if ($is_T(arg1)) {
+    var x2 = $as_T(arg1);
+    $$this$1.appendChild($m_Lorg_scalajs_dom_package$().document__Lorg_scalajs_dom_raw_HTMLDocument().createTextNode(x2))
+  } else if ($uZ((arg1 instanceof $g.Node))) {
+    $$this$1.appendChild(arg1)
+  } else if ($is_F1(arg1)) {
+    var x4 = $as_F1(arg1);
+    x4.apply__O__O($$this$1)
+  } else {
+    throw new $c_s_MatchError().init___O(arg1)
+  };
+  i = ((1 + i) | 0)
+};
+```
+
+Despite of the type safety issue due to the usage of `Any`, the above code are inefficient: 
+
+1. Unnecessary temporary object for the `xml.attributes.title(xml.text("my-title"))`.
+2. Unnecessary temporary `Seq` to hold repeated parameters.
+3. Unnecessary runtime type check for each argument.
+
+The similar issues can be found in many other usage of repeated parameters. For example, Scala string interpolation is inefficient due to its internal vararg function call, unless implementing it in a macro; Scala collection initializers (e.g. `List(1, 2, 3)`) create unnecessary temporary `Seq` before creating the desired collection.
+
+## Design
+
+This proposal introduces a new type `Curried` defined as following:
+
+``` scala
+trait Curried extends Any
+```
+
+When a function call `f(p1, p2, p3, ... pn)` is being type checked, the compiler will firstly look for `apply` method on `f`. If an applicable `apply` method is not found and `f` is a subtype of `Curried`, the compiler will convert the function call to curried form `f.applyBegin.applyNext(p1).applyNext(p2).applyNext(p3) ... .applyNext(pn).applyEnd`, and continue type checking the translated call.
```

Comment:
It is not clear from this spec how it interacts with the `scala.Dynamic` rewrites. Should it be tried before or after the `scala.Dynamic` rewrite? Should we prevent a class from extending both `scala.Dynamic` and `Curried`?